src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -63,12 +63,136 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
+                diagnostics.Add(ErrorCode.ERR_OperatorIsParameterCount, node.Location, type, subPatternsCount);
+                // PROTOTYPE(patterns): should we bind the subpatterns for the semantic model to use,
+                // PROTOTYPE(patterns): even though we don't have an expression to match them against?
+                return new BoundWildcardPattern(node, hasErrors: true);
+            }
+
+            // Note: we bind the pattern (conversion) to the first argument of `operator is`, not the type containing `operator is`.
+            // That is how we support the specification of pattern-matching external to a type (e.g. active patterns)
+            hasErrors = hasErrors || CheckValidPatternType(node.Type, operand, operandType, isOperator.Parameters[0].Type, false, diagnostics);
+            var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+            for (int i = 0; i < subPatternsCount; i++)
+            {
+                if (subPatterns[i].NameColon != null)
+                {
+                    // PROTOTYPE(patterns): keyword syntax not supported in recursive patterns.
+                    diagnostics.Add(ErrorCode.ERR_FeatureIsUnimplemented, subPatterns[i].NameColon.Location, "keyword syntax in recursive patterns");
+                    hasErrors = true;
+                }
+
+                patterns.Add(BindPattern(subPatterns[i].Pattern, null, isOperator.Parameters[i + 1].Type, hasErrors, diagnostics));
+            }
+
+            return new BoundRecursivePattern(node, type, isOperator, patterns.ToImmutableAndFree(), hasErrors);
+        }
+
+        /// <summary>
+        /// Is a user-defined `operator is` well-formed? At the use site, we ignore those that are not.
+        /// </summary>
+        private bool WellFormedOperatorIs(MethodSymbol candidate)
+        {
+            // requires at least one parameter
+            if (candidate.ParameterCount == 0) return false;
+
+            // must be static.
+            // PROTOTYPE(patterns): we don't support value-capturing matchers (e.g. regular expressions) yet.
+            if (!candidate.IsStatic) return false;
+
+            // the first parameter must be a value parameters. The remaining parameters must be out parameters.
+            foreach (var parameter in candidate.Parameters)
+            {
+                if (parameter.RefKind != ((parameter.Ordinal == 0) ? RefKind.None : RefKind.Out))
+                {
+                    return false;
+                }
+            }
+
+            // must return void or bool
+            switch (candidate.ReturnType.SpecialType)
+            {
+                case SpecialType.System_Void:
+                case SpecialType.System_Boolean:
+                    break;
+                default:
+                    return false;
+            }
+
+            // all constraints are satisfied
```

Comment:
There is no way to declare a generic operator in source, but I agree it should be filtered here. Accessibility too.